File: javalibrary/src/main/java/cn/yzl/startactivityforesult/java/StartActivityForResultHelperFragment.java
package cn.yzl.startactivityforesult.java;

import android.content.Intent;
import android.support.v4.app.Fragment;
import android.util.SparseArray;

public class StartActivityForResultHelperFragment extends Fragment {
    private final SparseArray<ActivityHelper.Callback> callbacks = new SparseArray();

    public final void startActivityForResult(Intent intent, int requestCode, ActivityHelper.Callback callback) {
        if (this.getActivity() != null) {
            this.callbacks.put(requestCode, callback);
            this.startActivityForResult(intent, requestCode);
        } else {
            throw new IllegalStateException("This fragment must be attached to an activity.");
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (!this.isDetached()) {
            ActivityHelper.Callback callback = this.callbacks.get(requestCode);
            if (callback != null) {
                ActivityResult activityResult = new ActivityResult(requestCode, resultCode, data);
                callback.onResult(activityResult);
            }
        }
    }

    public void onDetach() {
        super.onDetach();
        this.callbacks.clear();
    }
}

File: library/src/main/java/cn/yzl/startactivityforesult/library/StartActivityForResultHelperFragment.kt
package cn.yzl.startactivityforesult.library

import android.content.Intent
import android.support.v4.app.Fragment
import android.util.SparseArray

class StartActivityForResultHelperFragment : Fragment() {

    private val callbacks = SparseArray<(ActivityResult) -> Unit>()

    fun startActivityForResult(intent: Intent, requestCode: Int, callback: (ActivityResult) -> Unit) {
        val fragmentActivity = activity
                ?: throw IllegalStateException("This fragment must be attached to an activity.")
        callbacks.put(requestCode, callback)
        startActivityForResult(intent, requestCode)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (isDetached) {
            return
        }
        val callback = callbacks.get(requestCode) ?: return
        val activityResult = ActivityResult(requestCode, resultCode, data)
        callback(activityResult)
    }

    override fun onDetach() {
        super.onDetach()
        callbacks.clear()
    }
}
